src/server/master_server/master_metadata_service_impl.cc

```diff
@@ -159,7 +183,101 @@ grpc::Status MasterMetadataServiceImpl::HandleFileChunkRead(
+  for (auto chunk_server_location : reply->metadata().locations()) {
+    const std::string server_address(
+        chunk_server_location.server_hostname() + ":" +
+        std::to_string(chunk_server_location.server_port()));
+    // Create and return this chunk server Rpc client if not exist
+    auto chunk_server_service_client =
+        GetOrCreateChunkServerProtocolClient(server_address);
+    // Issue AdvanceFileChunkVersion request
+    AdvanceFileChunkVersionRequest advance_version_request;
+    advance_version_request.set_chunk_handle(chunk_handle);
+    // Advance the chunk version by 1
+    advance_version_request.set_new_chunk_version(chunk_version + 1);
+    grpc::ClientContext client_context;
+    common::SetClientContextDeadline(client_context, config_manager_);
+
+    // Issue AdvanceFileChunkVersion and check status
+    google::protobuf::util::StatusOr<AdvanceFileChunkVersionReply>
+        advance_version_reply_or(
+            chunk_server_service_client->SendRequest(advance_version_request,
+                                                     client_context));
+    if (!advance_version_reply_or.ok()) {
+      LOG(ERROR) << "Failed to advance chunk version for chunk " 
+                 << chunk_handle << " on chunk server " << server_address
+                 << "due to " << advance_version_reply_or.status();
+      // TODO(Xi): handle if all version advancement fails
+    } else {
+      LOG(INFO) << "Advanced chunk version for chunk " << chunk_handle
+                << " on chunk server " << server_address;
+    }
+  }
+
+  // After advancing the chunk version, obtain a lease
+  const std::string& primary_server_address(
+      reply->metadata().primary_location().server_hostname() + 
+      std::to_string(reply->metadata().primary_location().server_port()));
+  auto lease_service_client(GetOrCreateChunkServerProtocolClient(
+                                primary_server_address));
+  // Prepare GrantLeaseRequest to send to chunk server
+  GrantLeaseRequest grant_lease_request;
+  grant_lease_request.set_chunk_handle(chunk_handle);
+  grant_lease_request.set_chunk_version(chunk_version + 1);
+  grant_lease_request.mutable_lease_expiration_time()->set_seconds( 
+      absl::ToUnixSeconds(absl::Now() + 
+                              config_manager_->GetWriteLeaseTimeout()));
+  grpc::ClientContext client_context;
+  common::SetClientContextDeadline(client_context, config_manager_);
+
+  // Issue GrantLeaseRequest request and check status
+  google::protobuf::util::StatusOr<GrantLeaseReply> grant_lease_reply_or(
+      lease_service_client->SendRequest(grant_lease_request, client_context));
+
+  // Handle error, and logging
+  if (!grant_lease_reply_or.ok()) {
+    LOG(ERROR) << "Grant lease request for chunk " << chunk_handle 
+               << " at " << primary_server_address << " failed due to "
+               << grant_lease_reply_or.status();
+    return common::utils::ConvertProtobufStatusToGrpcStatus(
```

Comment:
technically, we can retry on another server if the random server cannot be a primary

Reply:
Yes, I added a TODO here (running out of time here).